### ISSUE TITLE
Help with a fork of the model that doesn't converge

### ISSUE THREAD

**Issue by jsomers** [potterhsu/SVHNClassifier#11]
Hi! Thanks very much for this. I was able to get the model to train on a Google Cloud machine in about ~15 hours.

I've created a fork (https://github.com/jsomers/ChessClassifier) whose goal is to recognizes images of chess boards, like this one:

![image](https://user-images.githubusercontent.com/21294/34642195-274211da-f2dd-11e7-9729-6fb226aa690b.png)

The idea is to say which pieces are in which square. So instead of the network terminating in 5 softmaxes of size 10, one for each of the digits, it terminates in 64 softmaxes of size 13 (13 because there are 6 possible black pieces, 6 possible white pieces, and the "null" piece, i.e., an empty square). I've removed the node that outputs a "length" because my length is always 64.

Unfortunately the model runs but doesn't train. The training accuracy is reported as "0.00000" and the loss hovers around ~60-65 indefinitely.

Is there any chance you might take a few minutes to discuss how I could get the model to converge? I have it on good authority that this problem is solvable, but perhaps my current model is broken in some way. My email address is listed on my website, which is listed on my Github profile. Thank you!

**Comment by MarStarck** [comment]
I think you don't need to train the whole image, split it into 8*8=64 and classify it into 13 classes separately would work fine.

**Comment by jsomers** [comment]
@MarStarck You're quite right, and that has been done successfully [here](https://github.com/Elucidation/tensorflow_chessbot), but in a way my whole goal was to avoid having to do that kind of preprocessing, both because it feels "cleaner" and because I think there's some useful constraint in having the whole board available to the model at once: two pawns of the same style and color seem more likely to be seen as such if they're viewed together, in the context of a single board, rather than apart, as individual pieces.

**Comment by jsomers** [comment]
@MarStarck @potterhsu With the help of a friend I was able to get the model to converge. Turns out I had a problem with exploding gradients (or perhaps it was vanishing gradients...). Here's the key change (in *model.py*):

```diff
-        train_op = optimizer.minimize(loss, global_step=global_step)
+        grad_vars = optimizer.compute_gradients(loss)
+        clipped_grad_vars = [(tf.clip_by_norm(grad, 1.), var) for grad, var in grad_vars]
+        train_op = optimizer.apply_gradients(clipped_grad_vars, global_step=global_step)
```

The commit is here: https://github.com/jsomers/ChessClassifier/commit/499348430bc1978ef3229584dc79049f781a4486

The key is to clip the gradients. Once I did that, the model actually started training, taking about 20 hrs or so. It now can find and recognize the pieces in grayscale chessboards like this one:

![image](https://user-images.githubusercontent.com/21294/36359159-0e6938c0-14e6-11e8-9d78-1d26aabf30ba.png)
